fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreReplaceIndexTest.java

```diff
@@ -0,0 +1,445 @@
+
+            assertTrue(recordStore.isIndexReadable(origIndex));
+            assertEquals(origEntries, scanIndex(origIndex));
+
+            disableIndex(newIndex1);
+            buildIndex(newIndex2);
+
+            commit(context);
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, withReplacementsHook);
+
+            assertTrue(recordStore.isIndexReadable(origIndex));
+            assertEquals(origEntries, scanIndex(origIndex));
+
+            buildIndex(newIndex1);
+            commit(context);
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, withReplacementsHook);
+            assertTrue(recordStore.isIndexDisabled(origIndex));
+            commit(context);
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, composeHooks(addAllIndexesHook, bumpMetaDataVersionHook(), bumpMetaDataVersionHook()));
+
+            assertTrue(uncheckedMarkIndexReadable(origIndex));
+            assertThat(scanIndex(origIndex), empty());
+        }
+    }
+
+    @Test
+    public void replacementIndexMissingInMetaDataFails() {
+        try (FDBRecordContext context = openContext()) {
+            MetaDataException err = assertThrows(MetaDataException.class, () -> openSimpleRecordStore(context, metaDataBuilder -> {
+                final Index index = new Index("indexWithFakeReplacement", Key.Expressions.field("num_value_2"),
+                        IndexTypes.VALUE,
+                        Collections.singletonMap(IndexOptions.REPLACED_BY_OPTION_PREFIX, "fakeIndex"));
+                metaDataBuilder.addIndex("MySimpleRecord", index);
+            }));
+            assertThat(err.getMessage(), containsString("Index indexWithFakeReplacement has replacement indexes [fakeIndex] that are not in the meta-data"));
+        }
+    }
+
+    @Test
+    public void replacementIndexPartiallyMissingInMetaDataFails() {
+        try (FDBRecordContext context = openContext()) {
+            MetaDataException err = assertThrows(MetaDataException.class, () -> openSimpleRecordStore(context, metaDataBuilder -> {
+                final Index index = new Index("indexWithOneFakeReplacement", Key.Expressions.field("num_value_2"),
+                        IndexTypes.VALUE,
+                        ImmutableMap.of(
+                                IndexOptions.REPLACED_BY_OPTION_PREFIX + "_00", "fakeIndex",
+                                IndexOptions.REPLACED_BY_OPTION_PREFIX + "_01", "MySimpleRecord$str_value_indexed"));
+                metaDataBuilder.addIndex("MySimpleRecord", index);
+            }));
+            assertThat(err.getMessage(), containsString("Index indexWithOneFakeReplacement has replacement indexes [fakeIndex] that are not in the meta-data"));
+        }
+    }
+
+    @Test
+    public void replacementIndexCycleFails() {
```

Comment:
Should there also be a test for `indexA` replacedBy `indexB, indexC` and `indexB` replaced by `indexD`, and `indexC` replaced by `indexD` (resulting in a non-directed cycle)

Reply:
Added that test, which fails validation now